Fix #9711 Failed to marshal HeaderWrapperView

humblehacker opened the pull request:
### Description of Change ###

Crash appears to be occurring due to the premature disposal of a `HeaderWrapperView` instance.

Crash occurs in `VisualElementRenderer::UpdateParentPageAccessibilityElements()`. As it walks up the superview chain from the `UILabel` that was tapped on in the test, it arrives at a superview whose `NSObject` cannot be retrieved. That superview happens to be a `HeaderWrapperView` instance. In `Runtime.cs:1249`, the call to `TryGetNSObject()` to retrieve this `HeaderWrapperView` returns `null`.  In the debugger, I manually called `TryGetNSObject()` with `true` for the `evenInFinalizerQueue` parameter, which returned the expected instance -- confirming that the instance has been unexpectedly garbage collected.

The resulting exception claims that it was not "possible to create a new managed instance (because the type 'Xamarin.Forms.Platform.iOS.HeaderWrapperView' does not have a constructor that takes one IntPtr argument)".  In my first attempted workaround, I added the missing constructors.  This got me past the initial problem, only to reveal another -- a call to `ShouldReceiveTouch()` with a `HeaderWrapperView` instance passed as the `UITouch` parameter.  That looked to me like a dangling pointer problem, so I decided to look elsewhere.

I'm very new to Xamarin/.NET, so someone with more context might be able to discover why this `HeaderWrapperView` instance was garbage collected. But I do know iOS, and when I looked into where and how `HeaderWrapperView` is used I saw some suspect code.  Typically, when implementing section headers with custom views in a `UITableView`, you want to use the same kind of reuse that you do for table view cells.  That is, you should use `UITableView.dequeueReusableHeaderFooterView()` to retrieve the section header view.  However, the current implementation creates a new `HeaderWrapperView` on each call to `ListViewDataSource.GetViewForHeader()`. In the test case, this happens each time the state changes in response to the tap on a section header.

To work around the problem, I used `UITableView.dequeueReusableHeaderFooterView()` to attempt reuse of `HeaderFooterView` so a new instance isn't created each time.  This appears to be sufficient to avoid the original issue.

**Additional considerations**

This header reuse only applies to `HeaderWrapperView`, and not to its subviews.  It would be better for scrolling performance if all subviews of `HeaderWrapperView` could be reused. I attempted to implement that (See commented out line at [`ListViewRenderer.cs:1122`](https://github.com/humblehacker/Xamarin.Forms/blob/3b8b1755e440cef4df352929d9597205b697bc16/Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs#L1122)), but was unable to get the subviews to refresh properly. Again someone with more Xamarin experience and context might be able to make this work.

### Issues Resolved ### 

- fixes #9711 

### API Changes ###

None

### Platforms Affected ### 

- iOS

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not Applicable

### Testing Procedure ###
<!-- Please list the steps that should be taken to properly test these changes on each relevant platform. If you were unable to test these changes yourself on any or all platforms, please let us know. Also, if you are able to attach a video of your test run, you will be our personal hero. -->

* Set project execution options to Xamarin.Forms.ControlGallery.iOS/Debug | iPhoneSimulator | iPhone SE (2nd generation)
* Build and run
* Open Unit Tests pad
* Run UITests → Xamarin.Forms.Core.iOS.UITests → Xamarin.Forms.Controls.Issues → Issue9711 → TestTappingHeaderDoesNotCrash

**Examples**

* [Test run before fix](https://www.icloud.com/iclouddrive/0CgGpTckHv9GazRUjWtHVF9dw#9711-before-fix)
* [Test run after fix](https://www.icloud.com/iclouddrive/0d_bSYkHEmGgixZ5TQVj0B6ew#9711-after-fix)

### PR Checklist ###
<!-- To be completed by reviewers -->

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=10759) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=10759) before we can accept your contribution.<br/><br/>:x: humblehacker [sign now](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=10759)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/Xamarin.Forms?pullRequest=10759) it.</sub>

rmarinho commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -1457,15 +1460,43 @@ void PreserveActivityIndicatorState(Element element)
 	{
+		public HeaderWrapperView(string reuseIdentifier) : base((NSString)reuseIdentifier)
+		{
+			Console.WriteLine($"Constructing HeaderWrapperView:{guid}");
```
we can drop the Console.WriteLine

rmarinho commented on Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs:
```diff
@@ -1114,11 +1113,15 @@ public override UIView GetViewForHeader(UITableView tableView, nint section)
-				view = new HeaderWrapperView { Cell = cell };
-				view.AddSubview(renderer.GetCell(cell, null, tableView));
+				header.TableViewCell = renderer.GetCell(cell, null, tableView);
+				// header.TableViewCell = renderer.GetCell(cell, header.TableViewCell, tableView);
```
Please remove commented code

humblehacker commented:
@rmarinho Thanks for your review!

I left the commented out code for your team to review the subview reuse issue I brought up under "Additional Considerations" in the PR description, with the intent that it would be removed before PR approval.  But as you requested, I've removed the comment and the leftover debugging code, and changed the description to link to the commented code in the previous commit.

Please let me know if there are any other changes you'd like to see.
